sw/device/tests/dif/dif_pwrmgr_smoketest.c

```diff
@@ -0,0 +1,80 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "sw/device/lib/base/mmio.h"
+#include "sw/device/lib/dif/dif_pwrmgr.h"
+#include "sw/device/lib/runtime/log.h"
+#include "sw/device/lib/testing/check.h"
+#include "sw/device/lib/testing/test_main.h"
+#include "sw/device/lib/usbdev.h"
+
+#include "hw/top_earlgrey/sw/autogen/top_earlgrey.h"  // Generated.
+
+static dif_pwrmgr_t pwrmgr;
+
+const test_config_t kTestConfig;
+const dif_pwrmgr_wakeup_reason_t test_wakeup_reason = {
+    .types = kDifPwrmgrWakeupTypeRequest,
+    .request_sources = kDifPwrmgrWakeupRequestSourceThree,
+};
+
+const dif_pwrmgr_wakeup_reason_t por_wakeup_reason = {
+    .types = 0,
+    .request_sources = 0,
+};
+
+static bool compare_wakeup_reasons(const dif_pwrmgr_wakeup_reason_t *lhs,
+                                   const dif_pwrmgr_wakeup_reason_t *rhs) {
+  return lhs->types == rhs->types &&
+         lhs->request_sources == rhs->request_sources;
+}
+
+bool test_main(void) {
+  bool test_passed = false;
+
+  CHECK(dif_pwrmgr_init(
+            (dif_pwrmgr_params_t){
+                .base_addr =
+                    mmio_region_from_addr(TOP_EARLGREY_PWRMGR_AON_BASE_ADDR),
+            },
+            &pwrmgr) == kDifPwrmgrOk);
+
+  // Assuming the chip hasn't slept yet, wakeup reason should be empty.
+  dif_pwrmgr_wakeup_reason_t wakeup_reason;
+  CHECK(dif_pwrmgr_wakeup_reason_get(&pwrmgr, &wakeup_reason) == kDifPwrmgrOk);
+
+  if (compare_wakeup_reasons(&wakeup_reason, &por_wakeup_reason)) {
+    LOG_INFO("Powered up for the first time, begin test");
+  } else if (compare_wakeup_reasons(&wakeup_reason, &test_wakeup_reason)) {
+    test_passed = true;
+    LOG_INFO("USB wakeup detected");
+  } else {
+    CHECK(false);
+  }
+
+  // Fake low power entry through usb
+  // Force usb to output suspend indication
+  if (!test_passed) {
+    usbdev_wake(true);
+    usbdev_force_suspend();
+    usbdev_force_dx_pullup(kDpSel, true);
+    usbdev_force_dx_pullup(kDnSel, true);
```

Comment:
Don't understand why both?

Reply:
o yeah see above, this is part of the hoakyness where I'm just doing something to ensure the signals will definitely be different.
This is why it's not really appropriate as a smoketest, since i'm relying on the outside environment to do something specific. 

I'll probably work on converting this to aon_timer.  But this quick little test just helped me verify the usb fsm does what I think it should on some level.

Reply:
I'll probably still keep this test in some form, specifically for DV.  Since enumerating usb is a huge pain for DV right now ... I just want to have something quick that'll be able to exercise the entry / exit behavior. 

I'll split this test out to something...maybe called dv_pwrmgr_usb_smoketest or something, and convert this smoketest to using something self contained like aon_timer.

Reply:
So if I got it correctly, you're actually forcing usbdev into suspend (meaning it won't wake up on its own if the D/N lines change), then you force the D/N lines such that the AON wake up logic is triggered. And because of the much slower AON clock, the system has enough time to enter low power mode before the wake up logic is triggered that then generates the interrupt?

Even if it's not perfect and it doesn't test the full behavior, this seems suitable for prototyping. As you say, enumeration takes a long time and you don't want to go through this every time you make one small change. But this is still better than nothing. It would be good to have that in the repo.

As a second test, we could maybe add such a low-power cycle to the hello_usbdev test in the Verilator simulation. There, we can't rely on the clock ratios but we have a DPI module that could trigger the wakeup. Plus we can also test the wakeup logic with different pinflip settings.

Reply:
Agree that this is a fine way to test the AON wake.

The verilator test will be a bit interesting because of the timescales involved to get an actual suspend detected.